File: 92. Reverse Linked List II.cpp
/*
Reverse a linked list from position m to n. Do it in-place and in one-pass.

For example:
Given 1->2->3->4->5->NULL, m = 2 and n = 4,

return 1->4->3->2->5->NULL.

Note:
Given m, n satisfy the following condition:
1 ≤ m ≤ n ≤ length of list.
*/


class Solution {
public:
	ListNode *reverseBetween(ListNode *head, int m, int n) {
	    if(m == n) return head;
		ListNode prehead(0);
		prehead.next = head;
		ListNode *pre = &prehead;
		while (--m)
		    pre = pre->next;
		ListNode *pstart = pre->next;
		n -= m;
		while (n--) {
		    ListNode *p = pstart->next;
		    pstart->next = p->next;
		    p->next = pre->next;
		    pre->next = p;
		}
		return prehead.next;
	}
};

File: Linked List/92. Reverse Linked List II.py
# Reverse a linked list from position m to n. Do it in one-pass.
#
# Note: 1 ≤ m ≤ n ≤ length of list.
#
# Example:
#
# Input: 1->2->3->4->5->NULL, m = 2, n = 4
# Output: 1->4->3->2->5->NULL

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# see: https://leetcode.com/problems/reverse-linked-list-ii/discuss/30666/Simple-Java-solution-with-clear-explanation

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        prev = dummy = ListNode(None)
        dummy.next = head

        for _ in range(m - 1):
            prev = prev.next
        start = prev.next
        then = start.next
        for _ in range(n - m):
            start.next = then.next
            then.next = prev.next
            prev.next = then
            then = start.next
        return dummy.next